numpy/f2py/crackfortran.py

```diff
@@ -2925,6 +2961,143 @@ def analyzevars(block):
+            if len(dimrange) == 1:
+                bound = param_parse(dimrange[0], params)
+                dimrange = range(1, int(bound)+1)
+            else:
+                lbound = param_parse(dimrange[0], params)
+                ubound = param_parse(dimrange[1], params)
+                dimrange = range(int(lbound), int(ubound)+1)
+        else:
+            outmess('param_eval: multidimensional array parameters not'
+                    ' supported: %s\n' % repr(n))
+
+        # Parse parameter value
+        v = (v[2:-2] if v.startswith('(/') else v).split(',')
+        v_eval = []
+        for item in v:
+            try:
+                item = eval(item, g_params, params)
+            except Exception as msg:
+                v_eval.append(item)
+                outmess('param_eval: got "%s" on %s\n' % (msg, repr(item)))
+            v_eval.append(item)
+
+        p = dict([])
+        for i in range(len(dimrange)):
+            p[dimrange[i]] = v_eval[i]
+    else:
+        try:
+            p = eval(v, g_params, params)
+        except Exception as msg:
+            p = v
+            outmess('param_eval: got "%s" on %s\n' % (msg, repr(v)))
+    return p
+
+
+def param_parse(d, params):
+    """Recursively parse array dimensions.
+
+    Parses the declaration of an array variable or parameter
+    `dimension` keyword, and is called recursively if the
+    dimension for this array is a previously defined parameter
+    (found in `params`).
+
+    Parameters
+    ----------
+    d : str
+        Fortran expression describing the dimension of an array.
+    params : dict
+        Previously parsed parameters declared in the Fortran source file.
+
+    Returns
+    -------
+    out : str
+        Parsed dimension expression.
+
+    Examples
+    --------
+
+    * If the line being analyzed is
+
+      `integer, parameter, dimension(2) :: pa = (/ 3, 5 /)`
+
+      then `d = 2` and we return immediately, with
+
+    >>> d = '2'
+    >>> param_parse(d, params)
+    2
+
+    * If the line being analyzed is
+
+      `integer, parameter, dimension(pa(1)) :: pb = (/1, 2, 3/)`
+
+      then `d = pa(1)`; since `pa` is a previously parsed parameter,
+      and `pa(1) = 3`, we call `param_parse` recursively, to obtain
+
+    >>> d = 'pa(1)'
+    >>> params = dict(pa={1: 3, 2: 5})
+    >>> param_parse(d, params)
+    3
+
+    """
+    if d.find("(") != -1:
+        dname = d[:d.find("(")]
+        ddims = d[d.find("(")+1:d.rfind(")")]
+    else:
+        return str(d)
+    if '(' in ddims:
+        # this dimension expression is an array element
+        if ddims[:ddims.find("(")] in params:
+            # this dimension expression is also a parameter;
+            # parse it recursively
+            index = int(param_parse(ddims, params))
+            return str(params[dname][index])
+    elif ddims in params:
+        # Reached the bottom of the recursion
+        index = params[ddims]
+        return str(params[dname][index])
```

Comment:
If the `if` is false here, this returns `None`, rather than the string its supposed to. Are you sure you want to do that?

I'll take another look at this later - I think the logic behind this function is still wrong.

A failing test case would be `      `integer, parameter, dimension(pa) :: pb = (/1, 2, 3/)`, where `pa = 3` - this returns `pa`, but should return `3`.

Reply:
Ok, I'll review this again and let you know when it's ready for your review, I don't want to waste your time if I can find that before you take a look at it. Thanks again!

Reply:
I think my complaint is the same as it was in this comment: https://github.com/numpy/numpy/pull/15457#discussion_r488167973. The algorithm I suggest there will pass the test case I give above.